### PR TITLE
sl-11: Refetch user after page refresh

### PR DESCRIPTION
**Pull request by kuslosolodka**
https://www.awesomescreenshot.com/video/15793570?key=ec3955a0ed47c77f53ec52c67b489104
unauthorized
get_current_user

### CODE REVIEW

File: backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts [hunk 1]
```diff
@@ -6,9 +8,13 @@ type DefaultApiHandlerOptions = {
...
 
+type HeadersInjected = {
+  headers: FastifyRequest['headers'];
+};
+
 type ApiHandlerOptions<
   T extends DefaultApiHandlerOptions = DefaultApiHandlerOptions,
-> = {
+> = HeadersInjected & {
   body: T['body'];
```
Comment (what1s1ove) [inline backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts]:
```suggestion
> = {
  headers: type
  body: T['body'];
```

Comment (what1s1ove) [reply backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts]: prefer `Parameters<ServerAppRouteParameters>`

File: backend/src/packages/auth/auth.controller.ts [hunk 2]
```diff
@@ -51,6 +51,11 @@ class AuthController extends Controller {
           }>,
         ),
     });
+    this.addRoute({
+      path: AuthApiPath.AUTH_USER,
+      method: 'GET',
+      handler: (options) => this.find(options),
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]:
```suggestion
      handler: (options) => this.getCurrent(options),
```

File: backend/src/packages/auth/auth.controller.ts [hunk 3]
```diff
@@ -136,6 +141,48 @@ class AuthController extends Controller {
       payload: await this.authService.signIn(options.body),
     };
   }
+
+  /**
+   * @swagger
+   * /auth/user:
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]:
```suggestion
   * /auth/current:
```

File: backend/src/packages/auth/auth.controller.ts [hunk 4]
```diff
@@ -136,6 +141,48 @@ class AuthController extends Controller {
...
+   *          content:
+   *            application/json:
+   *              schema:
+   *                type: object
+   *                properties:
+   *                  message:
+   *                    type: object
+   *                    $ref: '#/components/schemas/User'
+   */
+  private async find(options: ApiHandlerOptions): Promise<ApiHandlerResponse> {
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]:
```suggestion
  private async getCurrent(options: ApiHandlerOptions): Promise<ApiHandlerResponse> {
```

File: backend/src/packages/auth/auth.controller.ts [hunk 5]
```diff
@@ -136,6 +141,48 @@ class AuthController extends Controller {
...
+    const authHeader = options.headers.authorization;
+    const token = authHeader?.split(' ')[1];
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]:
```suggestion
    const [,token] = authHeader?.split(' ') ?? [];
```

File: backend/src/packages/auth/auth.controller.ts [hunk 6]
```diff
@@ -136,6 +141,48 @@ class AuthController extends Controller {
...
+    const user = await this.authService.findAuthUser(token as string);
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]: what if not a string?

File: backend/src/packages/auth/auth.service.ts [hunk 7]
```diff
@@ -52,6 +53,14 @@ class AuthService {
       user,
     };
   }
+
+  public async findAuthUser(token: string): Promise<UserAuthResponse> {
+    const decodedToken = this.tokenService.decode(token);
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.service.ts]: provide a generic

File: frontend/src/packages/auth/auth-api.ts [hunk 8]
```diff
@@ -35,6 +36,19 @@ class AuthApi extends HttpApi {
 
     return await response.json<UserSignUpResponseDto>();
   }
+
+  public async getUser(): Promise<UserAuthResponse> {
```
Comment (what1s1ove) [inline frontend/src/packages/auth/auth-api.ts]:
```suggestion
  public async getCurrent(): Promise<UserAuthResponse> {
```

File: frontend/src/pages/auth/auth.tsx [hunk 9]
```diff
@@ -27,6 +29,10 @@ const Auth: React.FC = () => {
     [dispatch],
   );
 
+  useEffect(() => {
+    void dispatch(userAuthActions.getAuthUser());
+  }, [dispatch]);
```
Comment (what1s1ove) [inline frontend/src/pages/auth/auth.tsx]: should be on all pages

File: frontend/src/slices/auth/auth.user.slice.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { createSlice } from '@reduxjs/toolkit';
```
Comment (what1s1ove) [inline frontend/src/slices/auth/auth.user.slice.ts]: you don't need an additional slice

File: shared/src/packages/auth/libs/enums/auth-api-path.enum.ts [hunk 11]
```diff
@@ -2,6 +2,7 @@ const AuthApiPath = {
   ROOT: '/',
   SIGN_UP: '/sign-up',
   SIGN_IN: '/sign-in',
+  AUTH_USER: '/user',
```
Comment (what1s1ove) [inline shared/src/packages/auth/libs/enums/auth-api-path.enum.ts]:
```suggestion
  AUTH_USER: '/current',
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts (hunk 1 above)
Comment (kuslosolodka) [reply backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts]: when I use headers as Parameters<ServerAppRouteParameters>, in mapRequest of controller package it's passing as IncomingHttpHeaders. I received this error in controller package: Property 'headers' is missing in type 'IncomingHttpHeaders' but required in type 'HeadersInjected'.

Comment (what1s1ove) [reply backend/src/libs/packages/controller/libs/types/api-handler-options.type.ts]: try this way

```ts
type DefaultApiHandlerOptions = {
  headers?: Parameters<ServerAppRouteParameters['handler']>[0]['headers'];
  body?: unknown;
  query?: unknown;
  params?: unknown;
  user?: UserAuthResponse | null;
};

type ApiHandlerOptions<
  T extends DefaultApiHandlerOptions = DefaultApiHandlerOptions,
> = {
  body: T['body'];
  query: T['query'];
  params: T['params'];
  headers: Parameters<ServerAppRouteParameters['handler']>[0]['headers'];
  user: T['user'];
};
```

File: backend/src/packages/auth/auth.controller.ts [hunk 12]
```diff
@@ -136,6 +141,58 @@ class AuthController extends Controller {
...
+   *                properties:
+   *                  message:
+   *                    type: object
+   *                    $ref: '#/components/schemas/User'
+   */
+  private async getCurrent(
+    options: ApiHandlerOptions,
+  ): Promise<ApiHandlerResponse> {
+    const authHeader = options.headers.authorization;
+    const [, token] = authHeader?.split(' ') ?? [];
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.controller.ts]:
```suggestion
    const [, token] = options.headers.authorization?.split(' ') ?? [];
```

File: shared/src/packages/auth/libs/enums/auth-api-path.enum.ts [hunk 13]
```diff
@@ -2,6 +2,7 @@ const AuthApiPath = {
   ROOT: '/',
   SIGN_UP: '/sign-up',
   SIGN_IN: '/sign-in',
+  AUTH_USER: '/current',
```
Comment (what1s1ove) [inline shared/src/packages/auth/libs/enums/auth-api-path.enum.ts]:
```suggestion
  CURRENT: '/current',
```

File: frontend/src/slices/auth/actions.ts [hunk 14]
```diff
@@ -20,4 +20,14 @@ const signUp = createAsyncThunk<
   return user;
 });
 
-export { signUp };
+const getAuthUser = createAsyncThunk<
```
Comment (what1s1ove) [inline frontend/src/slices/auth/actions.ts]:
```suggestion
const getCurrentUser = createAsyncThunk<
```

File: backend/src/packages/auth/auth.service.ts [hunk 15]
```diff
@@ -52,6 +53,14 @@ class AuthService {
       user,
     };
   }
+
+  public async findAuthUser<T>(token: string): Promise<UserAuthResponse> {
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.service.ts]:
```suggestion
  public async getCurrent(token: string): Promise<UserAuthResponse> {
```

File: backend/src/packages/auth/auth.service.ts [hunk 16]
```diff
@@ -52,6 +53,14 @@ class AuthService {
...
+    const decodedToken = this.tokenService.decode<T>(token);
+    const userId = decodedToken['userId'] as number;
+    const user = await this.userService.find(userId);
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.service.ts]:
```suggestion
    const { userId } = this.tokenService.decode<{ userId: number }>(token);
    const user = await this.userService.find(userId);
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/packages/auth/auth.service.ts [hunk 17]
```diff
@@ -52,6 +53,13 @@ class AuthService {
       user,
     };
   }
+
+  public async getCurrent(token: string): Promise<UserAuthResponse> {
+    const { userId } = this.tokenService.decode<{ userId: number }>(token);
+    const user = await this.userService.find(userId);
+
+    return user as UserAuthResponse;
```
Comment (what1s1ove) [inline backend/src/packages/auth/auth.service.ts]:
```suggestion
    return user;
```
handle if `userService.find` returns `null`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/pages/auth/auth.tsx [hunk 18]
```diff
@@ -27,6 +28,10 @@ const Auth: React.FC = () => {
     [dispatch],
   );
 
+  useEffect(() => {
+    void dispatch(authActions.getCurrentUser());
```
Comment (andanf-e) [inline frontend/src/pages/auth/auth.tsx]: Shouldn't we dispatch fetch current user only in single place?

File: frontend/src/libs/components/app/app.tsx [hunk 19]
```diff
@@ -15,6 +16,12 @@ const App: React.FC = () => {
     }
   }, [isRoot, dispatch]);
 
+  useEffect(() => {
+    if (isRoot) {
```
Comment (andanf-e) [inline frontend/src/libs/components/app/app.tsx]: Guess it shouldn't have a check if we are on root route, but have a check if there is no user in store

File: frontend/src/pages/auth/auth.tsx (hunk 18 above)
Comment (what1s1ove) [reply frontend/src/pages/auth/auth.tsx]: right, only in the app

File: frontend/src/libs/components/app/app.tsx (hunk 19 above)
Comment (what1s1ove) [reply frontend/src/libs/components/app/app.tsx]: right. add check in the action

**Review by andanf-e — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**